cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -711,6 +717,220 @@ class self_comparator {
+};
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class device_row_hasher {
+  friend class row_hasher;
+
+ public:
+  device_row_hasher() = delete;
+
+  __device__ auto operator()(size_type row_index) const noexcept
+  {
+    // Hash the first column w/ the seed
+    auto const initial_hash =
+      cudf::detail::hash_combine(hash_value_type{0},
+                                 type_dispatcher<dispatch_storage_type>(
+                                   _table.column(0).type(),
+                                   element_hasher_adapter<hash_function>{_has_nulls, _seed},
+                                   _table.column(0),
+                                   row_index));
+
+    auto it = detail::make_counting_transform_iterator(0, [=](size_type column_index) {
+      return cudf::type_dispatcher<dispatch_storage_type>(
+        _table.column(column_index).type(),
+        element_hasher_adapter<hash_function>{_has_nulls},
+        _table.column(column_index),
+        row_index);
+    });
+
+    // Hash each element and combine all the hash values together
+    // note that this starts at 1 and not 0 since we already hashed the first column
+    return detail::accumulate(
+      it + 1, it + _table.num_columns(), initial_hash, [](auto hash, auto h) {
+        return cudf::detail::hash_combine(hash, h);
+      });
+  }
+
+ private:
+  /**
+   * @brief Computes the hash value of an element in the given column.
+   *
+   * When the column is non-nested, this is a simple wrapper around the element_hasher.
+   * When the column is nested, this uses the element_hasher to hash the shape and values of the
+   * column.
+   */
+  template <template <typename> class hash_fn>
+  class element_hasher_adapter {
+   public:
+    __device__ element_hasher_adapter(
+      Nullate nulls,
+      uint32_t seed             = DEFAULT_HASH_SEED,
+      hash_value_type null_hash = std::numeric_limits<hash_value_type>::max()) noexcept
+      : _has_nulls(nulls), _seed(seed), _null_hash(null_hash)
+    {
+    }
+
+    template <typename T, CUDF_ENABLE_IF(not cudf::is_nested<T>())>
+    __device__ hash_value_type operator()(column_device_view const& col,
+                                          size_type row_index) const noexcept
+    {
+      return element_hasher<hash_fn, Nullate>(_has_nulls, _seed, _null_hash)
+        .template operator()<T>(col, row_index);
+    }
+
+    template <typename T, CUDF_ENABLE_IF(cudf::is_nested<T>())>
+    __device__ hash_value_type operator()(column_device_view const& col,
+                                          size_type row_index) const noexcept
+    {
+      auto hash                   = hash_value_type{_seed};
+      column_device_view curr_col = col.slice(row_index, 1);
```

Comment:
What is important to understand here is that we have limited the number of children to exactly one. If it is a struct, then we transform it into multiple columns, one for each leaf child. I would suggest reading the doc for decompose_structs function in row_operators.cu

Reply:
Oh, right! I had forgotten that logic. A comment indicating that "structs are assumed to be decomposed and contain only one child" would be extremely helpful here, since that precondition is not obvious.